### ISSUE TITLE
Use WixDependencyExtension in installer

### ISSUE THREAD

**Issue by felfert** [billziss-gh/winfsp#128]
I'm currently writing an installer for a FUSE app which will use WinFSP. In this installer, I would like to use the a WiX [Requires](http://wixtoolset.org/documentation/manual/v3/xsd/dependency/requires.html) element to make my WinFSP a required dependency of my product. However WinFSP's installer does not include a [Provides](http://wixtoolset.org/documentation/manual/v3/xsd/dependency/provides.html) element in it's Product.wxs.
So: Please use the WixDependencyExtension and provide a **well known* key which can be used by other installers (like mine) to make shure, that a compatible WinFSP is installed. I have prepared a small example [here](https://github.com/felfert/WixDependencyExample) so you can see what I mean.
Build is only tested on Linux, sorry. For building on Win, you need to rewrite the Makefile.

Thanks
 - Fritz
----

Before submitting this issue please review this checklist. Ideally all checkmarks should be checked upon submitting. (Use an x inside square brackets like so: [x])

- [x] **Issue type**: Please consider posting only bug reports or enhancement requests. Questions are better in the [WinFsp Google Group](https://groups.google.com/forum/#!forum/winfsp).
- [x] **No Duplicate**: Ensure that your issue has not been filed before. (Check open and closed issues.)
- [x] **Description**: Provide a descriptive title and a detailed explanation of the problem you are experiencing (for a bug report) or you are trying to solve (for an enhancement request).
- [N/A] **Reproduce**: For bug reports provide detailed information on how to reproduce the problem. For enhancement requests you do not need to provide this information, unless you find it relevant.
- [N/A] **Behaviors**: Provide information on the expected and actual behaviors.
- [N/A] **Environment**: For bug reports provide information about your OS version and build (e.g. 10.0.14393) and WinFsp version and build (e.g. 2017.2 or 1.2.17341). For enhancement requests you do not need to provide this information, unless you find it relevant.

**Comment by billziss-gh** [comment]
Let me start by saying that I am a complete Wix novice. So there may be many misunderstandings on how Wix and MSI works. I have some questions below:

- In SSHFS-Win I use the linked [technique](https://github.com/billziss-gh/sshfs-win/blob/master/sshfs-win.wxs#L39-L44) to check if a different MSI is already installed. Would this not be adequate for your purposes?

- Is the Wix dependency concept a native MSI concept or a Wix-only concept?

- Does Wix dependency provide feature like you cannot uninstall winfsp unless all dependent products are also uninstalled? This text in [Provides](http://wixtoolset.org/documentation/manual/v3/xsd/dependency/provides.html) suggests so:

    > This element is required for any product, feature, or bundle that will use the Dependency feature to properly reference count other products or features. It should be authored into a component that is always installed and removed with the product or features that contain it. This guarantees that product dependencies are not removed before those products that depend on them.

    - If I understand this correctly, this is certainly desirable.

- Does the Wix dependency extension require any external libraries?

    - I have tried very hard to keep WinFsp self-contained and without external dependencies. If you have an OS with version >= 7, it should just work without additional software. I would like things to be kept that way. If the Wix dependency extension suddenly requires .NET 4.5, I would not want this change.

- As I am not a Wix expert, I would welcome your help in adding this.

**Comment by felfert** [comment]
Q1: No, because it only protects installation of sshfs, not at removal of winfsp.

Q2: Yes, thats exactly what i does. It is however only a warning dialog. telling the user, that there are still other components depending on winfsp. The user can still decide to ignore that :-)

Q3: It is a WiX concept, implemented with standard MSI functionality. A provider (winfsp) basically creates a shared key with a well known name in the windows installer db. Then all other installers who decide to use this key must have the  corresponding Depends element using the same key name in some vital component of their .wxs. The resulting installer then looks up that key in the installer DB and increments its reference count during install. Optionally, it can verify, that the version of the Provider is in a certain range. Upon uninstall, it simply decrements the reference count. If an attempt to uninstall the provider is made while the reference count is > 0, it shows that specific warning dialog (or when running unattended, fails to uninstall)

Q4: AFAIK no. The [source of this extension](https://github.com/wixtoolset/wix3/blob/develop/src/ext/DependencyExtension/ca/wixdepca.cpp) is quite simple. There is a .Net part, but I believe, that is used during build time only. Anyway just try out my PR and if it requires .Net 4.5 then revert the changes (they are trivial).

Q5: PR coming up ...

**Comment by felfert** [comment]
Of course, this is only the provider part. The Depends part is very simple as well:

Just add the xmlns extension header and then add something like the following into some essential Component of your Installer. (A component which **always** gets installed)
```
<dep:Provides>
    <dep:Requires
        ProviderKey="WinFSPprovider"
        IncludeMinimum="yes"
        IncludeMaximum="yes"
        Minimum="1.2.17164"  <!-- Minimum version of WinFSP -->
        Maximum="x.x.x.x"  <!-- Whatever maximum version of WinFSP you want to depend on -->
    />
</dep:Provides>
```

All attributes except the ProviderKey are optional.

Finally, add the option _-ext WixDependencyExtension_ to the invocation of candle and light in yor Makefile, wixproj or whetever your build framework uses.

**Comment by felfert** [comment]
FYI:
I checked the answer for Q4 on a pristine win8.1 installation: No additional libs are required.

**Comment by billziss-gh** [comment]
@felfert thank you very much for the PR. I am hoping to look at it some time tomorrow.

**Comment by billziss-gh** [comment]
Fritz, thank you again for the explanation of how this feature works and your PR. I have now merged it into master and we can close this issue.

**Comment by billziss-gh** [comment]
@felfert 

BTW, I find it surprising that github insists on not listing you as a contributor to this project. According to this [link](https://help.github.com/articles/why-are-my-contributions-not-showing-up-on-my-profile/) it may be because you have not added the email address you use for git commits in your github profile.

You do not have to do this of course. I mention it because I had to double check that I did not do something stupid like merge your commits under my name. (I haven't as far as I can tell.)
